transaction/core/src/ring_signature/mlsag.rs

```diff
@@ -103,22 +121,22 @@ impl RingMLSAG {
     // * `check_value_is_preserved` - If true, check that the value of inputs equals
     //   value of outputs.
     // * `rng` - Randomness.
-    fn sign_with_balance_check<CSPRNG: RngCore + CryptoRng>(
+    fn sign_with_balance_check(
         message: &[u8],
-        ring: &[(CompressedRistrettoPublic, CompressedCommitment)],
+        ring: &[ReducedTxOut],
         real_index: usize,
         onetime_private_key: &RistrettoPrivate,
         value: u64,
         blinding: &Scalar,
         output_blinding: &Scalar,
         generator: &PedersenGens,
         check_value_is_preserved: bool,
-        rng: &mut CSPRNG,
-    ) -> Result<Self, Error> {
+        mut rng: &mut dyn CryptoRngCore,
```

Comment:
This part is a bit screwy, it seems like it should work with just `&mut dyn CryptoRngCore`.

However, when I had that, and `Scalar::from_random(rng)`, I get these build errors that seem odd to me:

```
   Compiling mc-transaction-core v1.3.0-pre0 (/tmp/mobilenode/transaction/core)
error[E0277]: the size for values of type `dyn mlsag::CryptoRngCore` cannot be known at compilation time
   --> transaction/core/src/ring_signature/mlsag.rs:170:39
    |
170 |             r[2 * i] = Scalar::random(rng);
    |                        -------------- ^^^ doesn't have a size known at compile-time
    |                        |
    |                        required by a bound introduced by this call
    |
    = help: the trait `Sized` is not implemented for `dyn mlsag::CryptoRngCore`
note: required by a bound in `Scalar::random`
   --> /tmp/mobilenode/cargo/git/checkouts/curve25519-dalek-a9a3f0edb196c00c/8791722/src/scalar.rs:568:19
    |
568 |     pub fn random<R: RngCore + CryptoRng>(rng: &mut R) -> Self {
    |                   ^ required by this bound in `Scalar::random`

error[E0277]: the size for values of type `dyn mlsag::CryptoRngCore` cannot be known at compilation time
   --> transaction/core/src/ring_signature/mlsag.rs:171:43
    |
171 |             r[2 * i + 1] = Scalar::random(rng);
    |                            -------------- ^^^ doesn't have a size known at compile-time
    |                            |
    |                            required by a bound introduced by this call
    |
    = help: the trait `Sized` is not implemented for `dyn mlsag::CryptoRngCore`
note: required by a bound in `Scalar::random`
   --> /tmp/mobilenode/cargo/git/checkouts/curve25519-dalek-a9a3f0edb196c00c/8791722/src/scalar.rs:568:19
    |
568 |     pub fn random<R: RngCore + CryptoRng>(rng: &mut R) -> Self {
    |                   ^ required by this bound in `Scalar::random`
```

So they are suggesting that dalek upstream should only require `RngCore + CryptoRng + ?Sized`. But obviously they don't and no one seems to require that. And logically it seems like they should be happy with a reference to a trait object anyways.

By making this `mut rng:...` and invoking dalek as `Scalar::random(&mut rng)`, I get it to compile.

There should be a more elegant way to convince rust to compile this though... if you have suggestions it would be helpful, gonna keep thinking on it

Reply:
hi @garbageslam! i suspect this is because `&dyn CryptoRng + RngCore` != `CryptoRng + RngCore`, if you wanted to do it the other way you'd need to use `pub trait RingSigner<RNG: CryptoRng + RngCore = SomeDefaultRng>`  and `fn ... (..., rng: &mut RNG, ...)`

Reply:
I was hoping not to make the trait itself generic, I was hoping to pass a &mut to trait object in because I think it may ease syntax, and FFI later.

It seems to me that `dyn X` generally implements X, and the error message is actually about `?Sized` and not `X` here. And indeed my code builds when I pass `&mut` to a `&mut dyn X`. It just seems that there should be a way to make this work without the second `mut` in there, maybe there's some other syntax I need to try...

Reply:
Digging further:

It looks that, in most places where we have a bound of `CryptoRng + RngCore`, it should usually be `+ ?Sized` as well, since in all most all places the callee's don't care if the object is sized. But that's a large ecosystem change at this point.

The `&mut &mut` code is compiling because in `rand_core`, they have auto-implemented the trait for references to the trait:

https://docs.rs/rand/latest/rand/trait.RngCore.html#impl-RngCore-for-%26%27a%20mut%20R
https://docs.rs/rand/latest/rand/trait.CryptoRng.html#impl-CryptoRng-for-%26%27a%20mut%20R

And the second `&mut` thing is automatically sized, because it is a reference.

So that I think is why this is compiling. 

I think this may possibly be the simplest way that this can work, I'm not sure there's another form of trait object we can pass that the compiler will accept here. Willing to be wrong

Reply:
I opened a PR here for Dalek upstream so that we could potentially simplify this in the future:

Reply:
Consider adding a comment linking to that PR or this thread.